plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,237 @@
+type Ctx = [Rank] -- the enclosing lambdas in scope that surround a let
+
+-- | During the first pass of the AST, we accumulate into a state the minimum rank of each lets encountered.
+type RankData = M.Map
+                 PLC.Unique -- ^ the identifier introduced by a let-binding
+                 ( Rank-- ^ the minimum rank for this let binding, i.e. where this identifier can topmost/highest float upwards to
+                 , PLC.Unique   -- ^ a principal name for this binding, used only because a let-Datatype-bind can introduce multiple identifiers
+                 )
+-- | During the second pass of the AST, we transform the 'RankData' to an assoc. list of depth=>lambda=>{let_unique} mappings left to process. This assoc. list is sorted by depth for easier code generation.
+type DepthData = [(Int,         -- ^ the depth
+                    M.Map       -- ^ a mapping of lambdanames belonging to this depth => letidentifiers to float
+                      PLC.Unique    -- ^ a lambda name
+                      (S.Set PLC.Unique)  -- ^ the let bindings that should be floated/placed under this lambda
+                  )
+                 ]
+
+-- | A simple table holding a let identifier to its RHS (a triple of annotation, recursivity, binding).
+type LetTable tyname name a = M.Map PLC.Unique (a, Recursivity, Binding tyname name a)
+
+-- | An arbitrary-minimal rank to signify that a let-binding has no lambda deps, and thus is allowed to float at the *toplevel* of the PIR program.
+topRank :: Rank
+topRank = ( 0                               -- ^ depth of the toplevel
+          , PLC.Unique { PLC.unUnique = -1} -- ^ a dummy unique
+          )
+
+-- | This function takes a 'Term', cleans the 'Term' from all its 'Let'-bindings and stores those lets into a separate table.
+-- Note: the output term is most-likely not a valid PIR program anymore.
+removeLets :: (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+            => Term tyname name a
+            -> (Term tyname name a, LetTable tyname name a)
+removeLets = flip runState M.empty . removeLets'
+ where removeLets' = \case
+         -- this overrides the 'termSubterms' functionality only for the 'Let' constructor
+         Let a r bs tIn -> do
+           bs' <- traverse (bindingSubterms removeLets') bs
+           let newMap = M.fromList $ fmap (\ b -> (bindingUnique b, (a,r,b))) bs'
+           modify (M.union newMap)
+           removeLets' tIn
+         t -> termSubterms removeLets' t
+
+-- | Traverses a Term to create a mapping of every let variable inside the term ==> to its corresponding rank.
+compRanks ::  (PLC.HasUnique (tyname a) PLC.TypeUnique, PLC.HasUnique (name a) PLC.TermUnique)
+              => Term tyname name a -> RankData
+compRanks pir = fst $ execRWS (compRanks' pir :: RWS Ctx () RankData ()) [] M.empty
+ where
+  addMaxRank freeVars n = do
+    lamRanks <- M.fromList . map (\ l -> (snd l, (l, snd l))) <$> ask
```

Comment:
You do this a few times, maybe worth a function. Also a bit incomprehensible :laughing: